src/vcBPA.cpp

```diff
@@ -792,97 +735,174 @@ size_t vcBPA_FindClosestTriangle(vcBPAGrid *pOldGrid, udDouble3 position)
-  itemData.oldGrid = vcBPAGrid::create(itemData.pGrid->center, minPos, maxPos);
+  if (pBuffer->pointCount > 0)
+  {
+    *ppNode = udAllocType(vcBPAOctNode, 1, udAF_Zero);
+    (*ppNode)->Init(center, extents);
 
-  udDouble3 aabbCenter = itemData.oldGrid.center;
-  udDouble3 aabbExtents = udDouble3::create(itemData.pManifold->gridSize / 2.0);
-  aabbExtents += udDouble3::create(2 * itemData.pManifold->ballRadius); // Add overlap
+    udDouble3 childSize = extents * .5;
+    if (childSize[pData->shortAxis] > pData->gridSize)
+    {
+      // If childSize is still bigger than gridSize, continue to descend
+      udSafeDeque_PushBack(pData->pLoadList, **ppNode);
+    }
+    else
+    {
+      udDouble3 minPos = center - extents;
+      udDouble3 maxPos = center + extents;
 
-  bool hasPoints = false;
-  bool hasNeighbours = false;
-  vcBPA_PopulateGrid(itemData.pManifold->pContext, itemData.pOldModel, nullptr, aabbCenter, aabbExtents, &itemData.oldGrid, &hasPoints, &hasNeighbours);
-  vcBPA_DoGrid(&itemData.oldGrid, itemData.pManifold->ballRadius);
+      for (int i = 0; i < 3; ++i)
+      {
+        // Adjust range to grab gridsize sections within a halfsize of min/max pos so 
+        // that we don't need to store grids and check if we've already added them
+        // This creates potential for errors on octree boundaries?
+        double minmod = udMod(minPos[i], pData->gridSize);
+        double maxmod = pData->gridSize - udMod(maxPos[i], pData->gridSize);
+        minPos[i] = minPos[i] - minmod + (minmod > pData->halfSize ? pData->gridSize : 0);
+        maxPos[i] = maxPos[i] + maxmod - (maxmod > pData->halfSize ? pData->gridSize : 0);
+      }
+
+      udDouble3 currPos;
+      for (currPos.x = minPos.x + pData->halfSize; currPos.x < maxPos.x; currPos.x += pData->gridSize)
+      {
+        for (currPos.y = minPos.y + pData->halfSize; currPos.y < maxPos.y; currPos.y += pData->gridSize)
+        {
+          for (currPos.z = minPos.z + pData->halfSize; currPos.z < maxPos.z; currPos.z += pData->gridSize)
+          {
+            vcBPAConvertItemData data = {};
+            data.pManifold = pData->pManifold;
+            if (vcBPA_AddGrid(pData, currPos, &data))
+              udSafeDeque_PushBack(pData->pConvertItemData, data);
+          }
+        }
+      }
+    }
+  }
 
-  udSafeDeque_PushBack(pData->pConvertItemData, itemData);
+  vdkPointBufferF64_Destroy(&pBuffer);
 }
 
 uint32_t vcBPA_GridGeneratorThread(void *pDataPtr)
 {
   vcBPAConvertItem *pData = (vcBPAConvertItem *)pDataPtr;
+  vcBPAOctNode node;
 
-  vcBPAGrid *pGrid = nullptr;
-  int i = 0;
+  while (pData->running)
+  {
+    if (udSafeDeque_PopBack(pData->pLoadList, &node) != udR_Success)
+    {
+      if (pData->pManifold->pointsCollected == pData->pManifold->targetPoints)
+        break;
 
-  vdkPointCloudHeader header;
-  vdkPointCloud_GetHeader(pData->pNewModel, &header);
+      udSleep(250);
```

Comment:
Is this necessary?

Reply:
Not entirely, but the chokepoint is memory blowout due to load in convertData, so this can also throttle that to an extent